src/main/java/io/supertokens/storageLayer/StorageLayer.java

```diff
@@ -560,4 +565,70 @@ public static List<List<TenantIdentifier>> getTenantsWithUniqueUserPoolId(Main m
         }
         return result;
     }
+
+    private static Storage[] getStoragesForApp(Main main, AppIdentifier appIdentifier) {
+        Map<String, Storage> userPoolToStorage = new HashMap<>();
+
+        Map<ResourceDistributor.KeyClass, ResourceDistributor.SingletonResource> resources =
+                main.getResourceDistributor()
+                        .getAllResourcesWithResourceKey(RESOURCE_KEY);
+        for (ResourceDistributor.KeyClass key : resources.keySet()) {
+            Storage storage = ((StorageLayer) resources.get(key)).storage;
+            if (key.getTenantIdentifier().toAppIdentifier().equals(appIdentifier)) {
+                userPoolToStorage.put(storage.getUserPoolId(), storage);
+            }
+        }
+        return userPoolToStorage.values().toArray(new Storage[0]);
+    }
+
+    public static TenantIdentifierStorageAndUserIdMapping getTenantIdentifierStorageAndUserIdMappingForUser(
+            Main main, TenantIdentifier tenantIdentifier, String userId, UserIdType userIdType)
+            throws StorageQueryException, TenantOrAppNotFoundException, UnknownUserIdException {
+        Storage storage = getStorage(tenantIdentifier, main);
+        tenantIdentifier.setStorage(storage);
+
+        UserIdMapping mapping = io.supertokens.useridmapping.UserIdMapping.getUserIdMapping(
+                tenantIdentifier.toAppIdentifier(), userId, userIdType);
+
+        if (mapping != null) {
+            return new TenantIdentifierStorageAndUserIdMapping(tenantIdentifier, storage, mapping);
+        }
+
+        if (((AuthRecipeStorage) storage).doesUserIdExist(tenantIdentifier.toAppIdentifier(), userId)) {
```

Comment:
you do not need to check this if the userIdType is external - just throw UnknownUserIdException